qiskit/primitives/containers/sampler_pub.py

```diff
@@ -0,0 +1,139 @@
+        self._parameter_values = parameter_values or BindingsArray()
+        self._shots = shots
+        self._shape = self._parameter_values.shape
+        if validate:
+            self.validate()
+
+    @property
+    def circuit(self) -> QuantumCircuit:
+        """A quantum circuit."""
+        return self._circuit
+
+    @property
+    def parameter_values(self) -> BindingsArray:
+        """A bindings array."""
+        return self._parameter_values
+
+    @property
+    def shots(self) -> int | None:
+        """An specific number of shots to run with (optional)."""
+        return self._shots
+
+    @classmethod
+    def coerce(cls, pub: SamplerPubLike, shots: int | None = None) -> SamplerPub:
+        """Coerce SamplerPubLike into SamplerPub.
+
+        Args:
+            pub: an object to be Sampler pub.
+            shots: an optional default number of shots to use if not
+                   already specified by the pub-like object.
+
+        Returns:
+            A coerced sampler pub.
+        """
+        if isinstance(pub, SamplerPub):
+            if pub.shots is None and shots is not None:
+                return cls(
+                    pub.circuit,
+                    pub.parameter_values,
+                    shots=shots,
+                    validate=False,
+                )
+            return pub
+        if isinstance(pub, QuantumCircuit):
+            return cls(circuit=pub, shots=shots)
+        if len(pub) not in [1, 2, 3]:
+            raise ValueError(
+                f"The length of pub must be 1, 2 or 3, but length {len(pub)} is given."
+            )
+        circuit = pub[0]
+        parameter_values = BindingsArray.coerce(pub[1]) if len(pub) > 1 else None
+        if len(pub) > 2 and pub[2] is not None:
+            shots = pub[2]
+        return cls(circuit=circuit, parameter_values=parameter_values, shots=shots, validate=False)
+
+    def validate(self):
+        """Validate the pub."""
+        if not isinstance(self.circuit, QuantumCircuit):
+            raise TypeError("circuit must be QuantumCircuit.")
+
+        self.parameter_values.validate()
+
+        if self.shots is not None:
+            if not isinstance(self.shots, Integral):
+                raise TypeError("shots must be an integer")
+            if self.shots < 0:
+                raise ValueError("shots must be non-negative")
+
+        # Cross validate circuits and parameter values
+        num_parameters = self.parameter_values.num_parameters
+        if num_parameters != self.circuit.num_parameters:
+            raise ValueError(
+                f"The number of values ({num_parameters}) does not match "
+                f"the number of parameters ({self.circuit.num_parameters}) for the circuit."
+            )
+
+
+SamplerPubLike = Union[
+    SamplerPub,
+    QuantumCircuit,
+    Tuple[QuantumCircuit],
+    Tuple[QuantumCircuit, BindingsArrayLike],
+    Tuple[QuantumCircuit, BindingsArrayLike, Integral],
```

Comment:
hmm, should we support this
```suggestion
    Tuple[QuantumCircuit, BindingsArrayLike | None, Integral],
```

Reply:
Technically i guess it can be `Tuple[QuantumCircuit, BindingsArrayLike | None, Integral | None]`